### PR TITLE
[core] add exponential backoff to task retry during oom

### PR DESCRIPTION
**Pull request by clarng**
Signed-off-by: Clarence Ng <clarence.wyng@gmail.com>

## Why are these changes needed?

Because the OOM killer kills task based on a pre-defined order, which is done to avoid live lock, it is possible that the same task gets re-submitted and killed, which may fail the task. Secondly, adding some form of retry delay will help produce back pressure and give time for longer running task to complete before it exhausts the retries.

Note this delay is only applied when the task fails due to OOM, and we maintain a separate delay for non-OOM retry that we may want to clean up (it is still used by a bunch of tests)

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/util/exponential_backoff_test.cc [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  // 2 ^ 80 will overflow.
+  auto backoff = ExponentialBackoff::GetBackoffMs(
+      /*attempt*/ 80,
+      /*base_ms*/ 1,
+      /*max_attempt*/ 50,
+      /*max_backoff_ms*/ std::numeric_limits<uint64_t>::max());
+  ASSERT_EQ(backoff, static_cast<uint64_t>(pow(2, 50)));
+}
+
+TEST(ExponentialBackoffTest, TestOverflowReturnsZero) {
```
Comment (scv119) [inline src/ray/util/exponential_backoff_test.cc:45]: hmm in this case we should not return 0 but instead the max_backoffs?

File: src/ray/core_worker/core_worker.h [hunk 2]
```diff
@@ -221,6 +221,28 @@ class TaskCounter {
   int64_t num_tasks_running_ GUARDED_BY(&mu_) = 0;
 };
 
+struct TaskToRetry {
+  /// Time when the task should be retried.
+  int64_t exection_time_ms;
```
Comment (scv119) [inline src/ray/core_worker/core_worker.h:226]: execution_time_ms ?

File: src/ray/common/ray_config_def.h [hunk 3]
```diff
@@ -396,6 +396,10 @@ RAY_CONFIG(uint64_t, global_gc_min_interval_s, 30)
 /// Duration to wait between retries for failed tasks.
 RAY_CONFIG(uint32_t, task_retry_delay_ms, 0)
 
+/// The base retry delay for exponential backoff when the task fails due to OOM.
+/// No delay if this value is zero.
+RAY_CONFIG(uint32_t, task_oom_retry_delay_base_ms, 1000)
```
Comment (scv119) [inline src/ray/common/ray_config_def.h]: disable by default?

**Review by scv119 — APPROVED** [review]

**Comment by clarng** [comment]
Test looks pretty much the same as master
PR: https://buildkite.com/ray-project/oss-ci-build-pr/builds/6025
maser around same time: https://buildkite.com/ray-project/oss-ci-build-branch/builds/1205

**Pull request merged**